### PR TITLE
Fix external app integrations for projects

### LINKED ISSUE #4703: Don't allow opening instances outside the current project with `EDIT` action
Covers behaviour documented at: https://docs.getodk.org/launch-collect-from-app/#using-a-uri-to-edit-a-form-or-instance. We need to make changes because of multi project setups.

- [x] `EDIT` for an instance with a URI without project ID works if the currently select project is the first (chronologically created) project
- [x] `EDIT` for an instance with a URI without a project ID shows a warning and closes if the currently select project is **NOT** the first (chronologically created) project
- [x] `EDIT` for an instance with a URI with a project ID works if the project is the currently selected one
- [x] `EDIT` for an instance with a URI with a project ID shows a warning and closes if the project is **NOT** the currently selected one

### LINKED ISSUE #4704: `PICK` action should return form and instance URIs with project ID
- [x] `PICK` for forms should return a URI with the `projectId` query parameter
- [x] `PICK` for instances should return a URI with the `projectId` query parameter

### PR DESCRIPTION
**Pull request by seadowg**
Closes #4703
Closes #4704

This ended up being pretty easy as forms and instances can very simply share the implementation for `EDIT` actions and it turns out `PICK` was already working due to us removing the old `CONTENT_URI` constants from the instances and forms contracts.

Interestingly, it looks like we don't actually support `PICK` for instances. Instead of fixing this, I propose we just amend our docs.

#### What has been done to verify that this works as intended?

New tests.

#### Why is this the best possible solution? Were any other approaches considered?

I'll leave comments inline on anything that stands out.

One thing to point out is there are two commits here dedicated to speeding up tests. I noticed that tests that added a project were running really slowly and found that, other than the disk write time, the slowness was caused by two things:

1. Waiting for the QR code scanner view to initialize
2. Repeatedly asking Android for the phone number while importing settings

It turns out that 1 can be fixed simply by making the view invisible in test (we stub out the decoder anyway) and that 2 could be solved by caching the phone number statically.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Again, we'll test this as part of general regression.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.

https://github.com/getodk/docs/issues/1375

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/androidTest/java/org/odk/collect/android/support/CollectTestRule.java [hunk 1]
```diff
@@ -67,11 +71,14 @@ public ShortcutsPage launchShortcuts() {
     }
 
     public <T extends Page<T>> T launch(Intent intent, T destination) {
-        /*
-        This can't use ActivityScenario.launch because of: https://github.com/android/android-test/issues/496
-         */
-        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
-        ApplicationProvider.getApplicationContext().startActivity(intent);
+        ActivityScenario.launch(intent);
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/CollectTestRule.java:73]: Looks like they fixed this.

File: collect_app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -188,15 +188,8 @@ the specific language governing permissions and limitations under the License.
...
-            <intent-filter>
-                <action android:name="android.intent.action.VIEW" />
-                <action android:name="android.intent.action.EDIT" />
-
-                <category android:name="android.intent.category.DEFAULT" />
-
-                <data android:mimeType="vnd.android.cursor.item/vnd.odk.instance" />
-            </intent-filter>
+            tools:ignore="AppLinkUrlError"
+            android:exported="true">
```
Comment (seadowg) [inline collect_app/src/main/AndroidManifest.xml:192]: We need `exported` here now that there isn't an intent filter.

**Comment by lognaturel** [comment]
>  it looks like we don't actually support PICK for instances

I'm quite confident this is a regression -- systems that wrote instances to disk as a means of communication with Collect used this. Also, [showing the instance list](https://docs.getodk.org/launch-collect-from-app/#launching-the-form-list-or-instance-list-activity) is useless without picking. Did you verify that it doesn't work in v1.30.1? This would have been tried with ODK Collect Tester so it would surprise me.

File: collect_app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -282,7 +275,7 @@ the specific language governing permissions and limitations under the License.
             </intent-filter>
         </activity>
 
-        <!-- Supports VIEW and EDIT actions for forms (documented at
+        <!-- Supports VIEW and EDIT actions for forms and instances (documented at
         https://docs.getodk.org/launch-collect-from-app/#using-a-uri-to-edit-a-form-or-instance).
         Do not use the activity directly as it's name/package might change. -->
```
Comment (lognaturel) [inline collect_app/src/main/AndroidManifest.xml]: `it's` to `its` while you're here? 🤓

**Comment by seadowg** [comment]
@lognaturel I could definitely be missing something, but the `AndroidManifest.xml` for 1.30.1 doesn't declare `PICK` action for instances (file [here](https://github.com/getodk/collect/blob/d539ef61b151ed66307cb40351dd40802c5afe36/collect_app/src/main/AndroidManifest.xml)). The `VIEW` and `EDIT` actions are supported for instances (from `InstanceChooserList`) but not `PICK`.

**Comment by seadowg** [comment]
@lognaturel it does look like the `InstanceChooserList` expects it (in the current version) though:

https://github.com/getodk/collect/blob/11ae4d29fa4c71c0777ffe20c61ccbd7fd52fc5b/collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java#L108

**Comment by seadowg** [comment]
@lognaturel ok digging even deeper it looks like this has never been supported but `InstanceChooserList` does seem to include the code that would make it work, so it's likely people are using the activity directly (which we don't want). I can add the `PICK` action here so people can migrate.

**Review by lognaturel — APPROVED** [review]

**Pull request merged**